source/common/event/scaled_range_timer.cc

```diff
@@ -0,0 +1,203 @@
+
+  void trigger() {
+    ASSERT(absl::holds_alternative<Active>(state_));
+    state_.emplace<Inactive>();
+    if (scope_ == nullptr) {
+      callback_();
+    } else {
+      ScopeTrackerScopeState scope(scope_, manager_.dispatcher_);
+      callback_();
+    }
+    scope_ = nullptr;
+  }
+
+private:
+  friend class ScaledRangeTimerManager;
+
+  struct Inactive {};
+
+  struct Pending {
+    Pending(MonotonicTime latest_trigger) : latest_trigger(latest_trigger) {}
+
+    const MonotonicTime latest_trigger;
+  };
+
+  struct Active {
+    Active(MonotonicTime::duration duration,
+           ScaledRangeTimerManager::BucketEnabledList::iterator bucket_position)
+        : duration(duration), bucket_position(std::move(bucket_position)) {}
+    const MonotonicTime::duration duration;
+    const ScaledRangeTimerManager::BucketEnabledList::iterator bucket_position;
+  };
+
+  void onPendingTimerComplete() {
+    ENVOY_LOG_MISC(info, "pending complete for {}", static_cast<void*>(this));
+    ASSERT(absl::holds_alternative<Pending>(state_));
+    Pending& pending = absl::get<Pending>(state_);
+
+    const MonotonicTime now = manager_.dispatcher_.timeSource().monotonicTime();
+    auto it = manager_.add(*this, pending.latest_trigger - now);
+    state_.emplace<Active>(pending.latest_trigger - now, it);
+  }
+
+  ScaledRangeTimerManager& manager_;
+  const TimerCb callback_;
+  const TimerPtr pending_timer_;
+
+  absl::variant<Inactive, Pending, Active> state_;
+  const ScopeTrackedObject* scope_;
+};
+
+ScaledRangeTimerManager::ScaledRangeTimerManager(Dispatcher& dispatcher, float scale_factor,
+                                                 std::chrono::milliseconds minimum_duration)
+    : dispatcher_(dispatcher), minimum_duration_(minimum_duration), scale_factor_(scale_factor) {}
+
+RangeTimerPtr ScaledRangeTimerManager::createTimer(TimerCb callback) {
+  return std::make_unique<ScaledRangeTimer>(callback, *this);
+}
+
+void ScaledRangeTimerManager::setScaleFactor(float scale_factor) {
+  scale_factor_ = DurationScaleFactor(scale_factor);
+  for (auto& bucket : buckets_) {
+    if (!bucket.scaled_timers.empty()) {
+      bucket.updateTimer(*this);
+    }
+  }
+}
+
+MonotonicTime::duration
+ScaledRangeTimerManager::getBucketedDuration(MonotonicTime::duration max_duration) {
+  return MonotonicTime::duration(static_cast<MonotonicTime::duration::rep>(
+      std::pow(kBucketScaleFactor,
+               std::floor(std::log(max_duration.count()) / std::log(kBucketScaleFactor)))));
+}
+
+void ScaledRangeTimerManager::Bucket::updateTimer(ScaledRangeTimerManager& manager) {
+  if (scaled_timers.empty()) {
+    if (timer->enabled()) {
+      timer->disableTimer();
+    }
+  } else {
+    auto& entry = scaled_timers.front();
+    timer->enableTimer(std::chrono::duration_cast<std::chrono::milliseconds>(
+        manager.scale_factor_.value() *
+        (entry.latest_trigger_time - manager.dispatcher_.timeSource().monotonicTime())));
```

Comment:
See comments about libevent monotonic time.  This will produce the wrong result, the timer will fire early.

Reply:
I don't think I understand enough about how libevent keeps time. You're saying `monotonicTime()` will be ahead of libevent's cached view of the current timestamp?

Reply:
I see, there's a Dispatcher::approximateMonotonicTime that I should be using instead.

Reply:
I don't think a correct API exists.  You may want to continue using manager.dispatcher_.timeSource().monotonicTime() until I get this fixed.